// @ngrx/router-store/esm2020/src/provide_router_store.mjs
import { ENVIRONMENT_INITIALIZER, inject } from '@angular/core';
import { _createRouterConfig, _ROUTER_CONFIG, ROUTER_CONFIG, } from './router_store_config';
import { FullRouterStateSerializer, } from './serializers/full_serializer';
import { MinimalRouterStateSerializer } from './serializers/minimal_serializer';
import { RouterStateSerializer, } from './serializers/base';
import { StoreRouterConnectingService } from './store_router_connecting.service';
/**
 * Connects the Angular Router to the Store.
 *
 * @usageNotes
 *
 * ```ts
 * bootstrapApplication(AppComponent, {
 *   providers: [
 *     provideStore({ router: routerReducer }),
 *     provideRouterStore(),
 *   ],
 * });
 * ```
 */
export function provideRouterStore(config = {}) {
    return {
        ɵproviders: [
            { provide: _ROUTER_CONFIG, useValue: config },
            {
                provide: ROUTER_CONFIG,
                useFactory: _createRouterConfig,
                deps: [_ROUTER_CONFIG],
            },
            {
                provide: RouterStateSerializer,
                useClass: config.serializer
                    ? config.serializer
                    : config.routerState === 0 /* RouterState.Full */
                        ? FullRouterStateSerializer
                        : MinimalRouterStateSerializer,
            },
            {
                provide: ENVIRONMENT_INITIALIZER,
                multi: true,
                useFactory() {
                    return () => inject(StoreRouterConnectingService);
                },
            },
            StoreRouterConnectingService,
        ],
    };
}
